nuclear-js/js/stores/CartStore.js

```diff
@@ -0,0 +1,53 @@
+import { Store, toImmutable } from 'nuclear-js'
+import {
+    CHECKOUT_START,
+    CHECKOUT_SUCCESS,
+    CHECKOUT_FAILED,
+    ADD_TO_CART,
+} from '../action-types'
+
+const initialState = toImmutable({
+    productQuantities: {},
+    pendingCheckout: {}
+})
+
+/**
+ * CartStores holds the mapping of productId => quantity
+ * and also maintains rollback information for the checkout process
+ */
+export default Store({
+    getInitialState() {
+        return initialState
+    },
+
+    initialize() {
+        this.on(CHECKOUT_START, beginCheckout)
+        this.on(CHECKOUT_SUCCESS, finishCheckout)
+        this.on(CHECKOUT_FAILED, rollback)
+        this.on(ADD_TO_CART, addToCart)
+    }
+})
+
+function addToCart(state, { product }) {
+    return (state.hasIn(['productQuantities', product.id]))
+        ? state.updateIn(['productQuantities', product.id], quantity => quantity + 1)
+        : state.setIn(['productQuantities', product.id], 1)
+}
+
+function beginCheckout(state) {
+    const currentItems = state.get('productQuantities')
+
+    return state
+        .set('productQuantities', toImmutable({}))
+        .set('pendingCheckout', currentItems)
+}
+
+function finishCheckout(state) {
+    return initialState
```

Comment:
no this should be `initialState` as its taking the `currentState` and returning the original state since the `finishCheckout` function clears the cart items.

I tend to leave the `state` parameter in store handler functions to show the transformation that is happening.